### ISSUE TITLE
`nlimbs!()` macro doesn't support rounding up to the limb size

### ISSUE THREAD

**Issue by BrettWitty** [RustCrypto/crypto-bigint#258]
The `nlimbs!` macro assumes the bit-size provided is a multiple of the limb bit-size. This is reflected in all the tests for the macro and its documentation.

This falls down for examples like `nlimbs!(63) = 0` (should be 1), and `nlimbs!(65) = 1` (should be 2) on a 64-bit platform.

**Comment by tarcieri** [comment]
The macro (and in general, the library as a whole) is not really designed to handle numbers that aren’t multiples of 32.

It should probably generate a compile error if invoked with such a number.

**Comment by BrettWitty** [comment]
That's a shame. There's circumstances in cryptography where you have goofy bit-sizes, even though you know the actual CPU arithmetic is on whole words.

I guess the library only ensures it fits in whole limbs, not part ones, which might add overhead to maintain.

**Comment by tarcieri** [comment]
Generally the approach there is to use the minimum number of limbs which can represent the value in question

**Comment by BrettWitty** [comment]
Yeah, that's what I was hoping from `nlimbs!()`.

**Comment by tarcieri** [comment]
Can I ask why you're using the macro at all as opposed to working with one of the type aliases?

Due to a lack of `feature(generic_const_exprs)` much of the functionality is only defined for specific sized integers.

**Comment by BrettWitty** [comment]
I wanted to be explicit about the type with a custom `Uint`, and create it like all the built ins. Of course `Uint< nlimbs!(127)>` should be the exact same type as `U128` if `nlimbs!` was calculated correctly, but is slightly more explicit in intent of what you want to create.

I can get around it with just defining `U127 = U128`, but it looks odd. No big deal, but your documentation for `nlimbs!` is a little misleading because it does not state its assumptions. Clarifying the assumption in doco would be sufficient to deal with weird crate users like me.

**Comment by tarcieri** [comment]
The problem is `U127` is not `U128`. It's a distinct type we don't support.

Supporting something like `U127` would require a newtype which handles masking out (or carry propagating, where relevant) the high bit. Sort of like `Residue`, but with an even modulus.

While `nlimbs!` could give you a "big enough" number, it's still not specifically going to be the one you asked for.

**Comment by BrettWitty** [comment]
Absolutely agree, and it's a fair point. I guess it would help to make this explicit in the docs.

**Comment by matzayonc** [comment]
In my opinion current behavior of this macro is confusing and will lead to errors. In my opinion it should round up like so:
```
#[macro_export]
macro_rules! nlimbs {
    ($bits:expr) => {
        $bits + $crate::Limb::BITS - 1 / $crate::Limb::BITS
    };
}
```
I would gladly create a PR if you agree.

**Comment by tarcieri** [comment]
@matzayonc did you read any of the previous discussion?

The problem with a "round up" strategy is `nlimbs!` for a non-multiple of the limb size won't give you an integer with that number of bits, but rather the next largest integer with that number of limbs, which I think is confusing.

I think it would be less confusing to generate errors in that case.

**Comment by matzayonc** [comment]
@tarcieri I did read it I do see your point. Error would definitely be better than current implementation.

I just think it would be a good idea to implement `nclosestlimbs!` that would work by rounding up.
My use case is writing tests for generic functions with smaller key (so nlimbs) sizes like 16. I've implemented it manually but I think it would be a useful as a part of the crate. 

With that I can use generic interface and allow for any input like 
   `pub n: Uint<{ (L + BITS - 1) / BITS }>,`
